src/mlpack/methods/ann/layer/mean_pooling_impl.hpp

```diff
@@ -0,0 +1,201 @@
+{
+  if (&other != this)
+  {
+    Layer<MatType>::operator=(other);
+    kernelWidth = other.kernelWidth;
+    kernelHeight = other.kernelHeight;
+    strideWidth = other.strideWidth;
+    strideHeight = other.strideHeight;
+    floor = other.floor;
+    channels = other.channels;
+    pooling = other.pooling;
+  }
+
+  return *this;
+}
+
+template<typename MatType>
+MeanPoolingType<MatType>&
+MeanPoolingType<MatType>::operator=(MeanPoolingType&& other)
+{
+  if (&other != this)
+  {
+    Layer<MatType>::operator=(std::move(other));
+    kernelWidth = std::move(other.kernelWidth);
+    kernelHeight = std::move(other.kernelHeight);
+    strideWidth = std::move(other.strideWidth);
+    strideHeight = std::move(other.strideHeight);
+    floor = std::move(other.floor);
+    channels = std::move(other.channels);
+    pooling = std::move(other.pooling);
+  }
+
+  return *this;
+}
+
+template<typename MatType>
+void MeanPoolingType<MatType>::Forward(
+    const MatType& input, MatType& output)
+{
+  arma::Cube<typename MatType::elem_type> inputTemp(
+      const_cast<MatType&>(input).memptr(), this->inputDimensions[0],
+      this->inputDimensions[1], input.n_cols * channels, false, false);
+
+  arma::Cube<typename MatType::elem_type> outputTemp(output.memptr(),
+      this->outputDimensions[0], this->outputDimensions[1],
+      input.n_cols * channels, false, true);
+
+  PoolingOperation(inputTemp, outputTemp);
+}
+
+template<typename MatType>
+void MeanPoolingType<MatType>::Backward(
+  const MatType& input,
+  const MatType& gy,
+  MatType& g)
+{
+  arma::Cube<typename MatType::elem_type> mappedError =
+      arma::Cube<typename MatType::elem_type>(((MatType&) gy).memptr(),
+      this->outputDimensions[0], this->outputDimensions[1],
+      channels * input.n_cols, false, false);
+
+  arma::Cube<typename MatType::elem_type> gTemp(g.memptr(),
+      this->inputDimensions[0], this->inputDimensions[1],
+      channels * input.n_cols, false, true);
+
+  gTemp.zeros();
+  for (size_t s = 0; s < mappedError.n_slices; s++)
+  {
+    Unpooling(mappedError.slice(s), gTemp.slice(s));
+  }
+}
+
+template<typename MatType>
+void MeanPoolingType<MatType>::ComputeOutputDimensions()
+{
+  this->outputDimensions = this->inputDimensions;
+
+  // Compute the size of the output.
+  if (floor)
+  {
+    this->outputDimensions[0] = std::floor((this->inputDimensions[0] -
+        (double) kernelWidth) / (double) strideWidth + 1);
+    this->outputDimensions[1] = std::floor((this->inputDimensions[1] -
+        (double) kernelHeight) / (double) strideHeight + 1);
+  }
+  else
+  {
+    this->outputDimensions[0] = std::ceil((this->inputDimensions[0] -
+        (double) kernelWidth) / (double) strideWidth + 1);
+    this->outputDimensions[1] = std::ceil((this->inputDimensions[1] -
+        (double) kernelHeight) / (double) strideHeight + 1);
+  }
+
+  // Higher dimensions are not modified.
+
+  // Cache input size and output size.
```

Comment:
Doesn't this just cache the total number of higher-dimension points?  (I am not sure I stated that clearly but I think you know what I mean.)

Reply:
Done.